### PR TITLE
Add support for array and traversable

### LINKED ISSUE #31: Handle type array in the DoctrineORMQuerySourceIterator
Hi,

i'm playing with Sonata Admin Bundle, the export feature use this library to output a file  in different formats : json,csv . I saw that "array type" for a doctrine entity is handled by returning $value=null. ([line105 DoctrineORMQuerySourceIterator.php](https://github.com/sonata-project/exporter/blob/master/lib/Exporter/Source/DoctrineORMQuerySourceIterator.php#L105)). So the json exported of an entity that have fields/field of "type array" is not complete. I want to ask why we can't return the array?

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because BC.

Closes #31.

## Changelog

```markdown
### Added
- Added support for array and traversable in `DoctrineORMQuerySourceIterator`, `DoctrineODMQuerySourceIterator` and `PropelCollectionSourceIterator`.
```

### CODE REVIEW

File: src/Source/DoctrineORMQuerySourceIterator.php [hunk 1]
```diff
@@ -13,55 +13,21 @@
...
 use Doctrine\ORM\Query;
 use Sonata\Exporter\Exception\InvalidMethodCallException;
-use Symfony\Component\PropertyAccess\Exception\UnexpectedTypeException;
-use Symfony\Component\PropertyAccess\PropertyAccess;
-use Symfony\Component\PropertyAccess\PropertyAccessor;
-use Symfony\Component\PropertyAccess\PropertyPath;
 
-class DoctrineORMQuerySourceIterator implements SourceIteratorInterface
+/**
+ * @final since sonata/exporter 2.x
```
Comment (phansys) [inline src/Source/DoctrineORMQuerySourceIterator.php:20]:
```suggestion
 * @final since sonata-project/exporter 2.x.
```

File: tests/Source/DoctrineORMQuerySourceIteratorTest.php [hunk 2]
```diff
@@ -52,8 +52,8 @@ public function getValueProvider()
         $dateTimeImmutable = new \DateTimeImmutable();
 
         $data = [
-            [[1, 2, 3], null],
-            [new \ArrayIterator([1, 2, 3]), null],
+            [[1, 2, 3], '[1, 2, 3]'],
+            [new \ArrayIterator([1, 2, 3]), '[1, 2, 3]'],
```
Comment (phansys) [inline tests/Source/DoctrineORMQuerySourceIteratorTest.php:56]:
```suggestion
            [new \ArrayIterator([1, 2, 3]), '[1, 2, 3]'],
            [(static function () { yield from [1, 2, 3]; })(), '[1, 2, 3]'],
```

File: tests/Source/DoctrineORMQuerySourceIteratorTest.php [hunk 3]
```diff
@@ -52,8 +52,8 @@ public function getValueProvider()
         $dateTimeImmutable = new \DateTimeImmutable();
 
         $data = [
-            [[1, 2, 3], null],
-            [new \ArrayIterator([1, 2, 3]), null],
```
Comment (phansys) [inline tests/Source/DoctrineORMQuerySourceIteratorTest.php:56]: Why did you remove this case?

Comment (VincentLanglet) [reply tests/Source/DoctrineORMQuerySourceIteratorTest.php:56]: Oh, miss read your comment.

File: src/Source/AbstractPropertySourceIterator.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    protected function getCurrentData($current): array
+    {
+        $data = [];
+        foreach ($this->fields as $key => $field) {
+            if (\is_string($field)) {
+                $name = \is_string($key) ? $key : $field;
+                $propertyPath = $field;
+            } else {
+                throw new \TypeError('Unsupported field type. Field should be a string.');
+            }
```
Comment (phansys) [inline src/Source/AbstractPropertySourceIterator.php:140]:
```suggestion
            if (!\is_string($field)) {
                throw new \TypeError(sprintf(
                    'Field type must be string, %s given.',
                    \is_object($field) ? 'instance of '.\get_class($field) : \gettype($field)
                );
            }

            $name = \is_string($key) ? $key : $field;
            $propertyPath = $field;
```

Comment (VincentLanglet) [reply src/Source/AbstractPropertySourceIterator.php:140]: I plan to support another type https://github.com/sonata-project/exporter/pull/341

Comment (phansys) [reply src/Source/AbstractPropertySourceIterator.php:140]: Great. IMO, if we move the exception before there is no need for the `else` clause. When the new type gets support, an `or` condition could be added (`if (!\is_string($field) && !$field instanceof ExportFieldInterface)`).

Comment (VincentLanglet) [reply src/Source/AbstractPropertySourceIterator.php:140]: This will be
```
if (\is_string($field) {
    doThings();
else ($field instanceof ExportFieldInterface) {
    doOtherThings();
} else {
    throw new Exception();
}
```

File: src/Source/AbstractPropertySourceIterator.php [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    protected $propertyAccessor;
+
+    /**
+     * @var string default DateTime format
+     */
+    protected $dateTimeFormat;
+
+    /**
+     * @var array<string>
```
Comment (greg0ire) [inline src/Source/AbstractPropertySourceIterator.php:51]: IMO we should either use

```suggestion
     * @var string[]
     * @psalm-var list<string>
```

or 

```suggestion
     * @var list<string>
```

or

```suggestion
     * @var string[]
```

File: src/Source/AbstractPropertySourceIterator.php [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+        if ('' === $datePart && '' === $timePart) {
+            return 'P0Y';
+        }
+
+        return 'P'.$datePart.('' !== $timePart ? 'T'.$timePart : '');
+    }
+
+    /**
+     * @param object|array $current
```
Comment (greg0ire) [inline src/Source/AbstractPropertySourceIterator.php:129]: array of what? `mixed`?

File: src/Source/AbstractPropertySourceIterator.php [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param mixed $value
+     *
+     * @return string|null
```
Comment (greg0ire) [inline src/Source/AbstractPropertySourceIterator.php:158]: Why is this not a return type hint?

File: src/Source/AbstractPropertySourceIterator.php (hunk 6 above)
Comment (VincentLanglet) [reply src/Source/AbstractPropertySourceIterator.php:129]: I don't know, so yeah mixed maybe. I took the typehint from property accessor getValue.

File: src/Source/AbstractPropertySourceIterator.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Source/AbstractPropertySourceIterator.php:158]: To be BC. The method was already existing.

File: src/Source/AbstractPropertySourceIterator.php (hunk 5 above)
Comment (VincentLanglet) [reply src/Source/AbstractPropertySourceIterator.php:51]: I wanted to introduce an ExportField type, so I would change to `@var array<string|ExportField>`.
But I can use `string[]` atm if you prefer.

**Comment by VincentLanglet** [comment]
Phpdoc is updated @greg0ire :)

File: src/Source/AbstractPropertySourceIterator.php [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+        if ('' === $datePart && '' === $timePart) {
+            return 'P0Y';
+        }
+
+        return 'P'.$datePart.('' !== $timePart ? 'T'.$timePart : '');
+    }
+
+    /**
+     * @param object|array<mixed> $current
```
Comment (greg0ire) [inline src/Source/AbstractPropertySourceIterator.php:129]:
```suggestion
     * @param object|mixed[] $current
```

Comment (VincentLanglet) [reply src/Source/AbstractPropertySourceIterator.php]: Oops!...I Did It Again

Fixed !
